packages/core/src/defer/idle_service.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {AbstractType} from '../interface/type';
+import {InjectionToken} from '../di/injection_token';
+import type {EnvironmentProviders} from '../di/interface/provider';
+import {makeEnvironmentProviders} from '../di/provider_collection';
+
+/**
+ * Use shims for the `requestIdleCallback` and `cancelIdleCallback` functions for
+ * environments where those functions are not available (e.g. Node.js and Safari).
+ *
+ * Note: we wrap the `requestIdleCallback` call into a function, so that it can be
+ * overridden/mocked in test environment and picked up by the runtime code.
+ */
+const _requestIdleCallback = () =>
+  typeof requestIdleCallback !== 'undefined' ? requestIdleCallback : setTimeout;
+const _cancelIdleCallback = () =>
+  typeof requestIdleCallback !== 'undefined' ? cancelIdleCallback : clearTimeout;
+
+/**
+ * Service which configures custom 'on idle' behavior for Angular features like `@defer`.
+ *
+ * @publicApi
+ */
+export interface IdleService {
+  /**
+   * Schedule `callback` to be executed when the current application or browser is considered idle.
+   *
+   * @returns an id which allows the scheduled callback to be cancelled before it executes.
+   */
+  requestOnIdle(callback: () => void): number;
```

Comment:
WDYT of having this return a cancellation callback, dropping `cancelOnIdle`:

```suggestion
  requestOnIdle(callback: () => void): () => void;
```